src/olympia/versions/models.py

```diff
@@ -202,6 +203,21 @@ def from_upload(cls, upload, addon, platforms, channel, send_signal=True,
         if send_signal:
             version_uploaded.send(sender=version)
 
+        # Generate a preview and icon for listed static themes
+        if (addon.type == amo.ADDON_STATICTHEME and
+                channel == amo.RELEASE_CHANNEL_LISTED):
+            from olympia.addons.models import Preview  # Avoid circular ref.
+            dst_root = os.path.join(user_media_path('addons'), str(addon.id))
+            theme_data = parsed_data.get('theme', {})
+            version_root = os.path.join(dst_root, unicode(version.id))
+
+            utils.extract_header_img(
+                version.all_files[0].file_path, theme_data, version_root)
+            preview = Preview.objects.create(
+                addon=addon, caption=unicode(version.version))
+            generate_static_theme_preview(
```

Comment:
This should be called with `.delay()` otherwise it's going to be synchronous...

Reply:
Can we have a task be done as priority?  (or only in `CELERY_TASK_ROUTES`?)  I'd like the preview to be finished by the time the user gets to the finish step if possible to display it to them....
Actually, ideally I'd block at that stage (though I don't want to block just here)

Reply:
I added it to `CELERY_TASK_ROUTES` in https://github.com/mozilla/addons-server/pull/7475/commits/86844e3fbdf9614a69824787c6e4d3ca64a7bf8a